File: DecodeWays.java
// https://leetcode.com/problems/decode-ways/
// 1D DP Recursive sol
class Solution {
    Map<Integer, Integer> memo = new HashMap<>();
    public int numDecodings(String s) {
        return recursiveWithMemo(0, s);
    }
    private int recursiveWithMemo(int index, String str) {
        // Have we already seen this substring?
        if (memo.containsKey(index)) {
            return memo.get(index);
        }
        // If you reach the end of the string
        // Return 1 for success.
        if (index == str.length()) {
            return 1;
        }
        // If the string starts with a zero, it can't be decoded
        if (str.charAt(index) == '0') {
            return 0;
        }
        if (index == str.length() - 1) {
            return 1;
        }
        int ans = recursiveWithMemo(index + 1, str);
        if (Integer.parseInt(str.substring(index, index + 2)) <= 26) {
             ans += recursiveWithMemo(index + 2, str);
         }
        memo.put(index, ans);
        return ans;
    }
}

File: DecodeWays.py
# https://leetcode.com/problems/decode-ways/
# 1D DP, ITERATIVE APPROACH
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == "0":
            return 0
        twoBack = 1
        oneBack = 1
        for i in range(1, len(s)):
            current = 0
            if s[i] != "0":
                current = oneBack
            twoDigit = int(s[i - 1: i + 1])
            if twoDigit >= 10 and twoDigit <= 26:
                current += twoBack
            twoBack = oneBack
            oneBack = current
        return oneBack